onnx/checker.cc

```diff
@@ -304,6 +304,133 @@ void check_attribute(
   }
 }
 
+void verify_node_schema(
+    const NodeProto& node,
+    const OpSchema* schema,
+    int domain_version) {
+  if (!schema ) {
+    fail_check(
+        "No Op registered for " + node.op_type() + " with domain_version of " +
+        ONNX_NAMESPACE::to_string(domain_version));
+  } else if (schema->Deprecated()) {
+    fail_check(
+        "Op registered for " + node.op_type() + " is deprecated in domain_version of " +
+        ONNX_NAMESPACE::to_string(domain_version));
+  }  else {
+    schema->Verify(node);
+  }
+}
+
+void check_function_node(
+    const NodeProto& node,
+    const CheckerContext& ctx,
+    const LexicalScopeContext& lex_ctx,
+    OpName_Domain_Version_Schema_Map& map_,
+    TENSOR_TYPES_MAP& tc_map) {
+  enforce_non_empty_field(node, op_type);
+
+  if (node.input().empty() && node.output().empty()) {
+    fail_check(
+        "NodeProto (name: ",
+        node.name(),
+        ", type: ",
+        node.op_type(),
+        ") has zero input and zero output.");
+  }
+
+  // Resolve domain for node
+  const auto& opset_imports = ctx.get_opset_imports();
+  auto dit = opset_imports.find(node.domain());
+  if (dit == opset_imports.end()) {
+    fail_check("No opset import for domain '" + node.domain() + "'");
+  }
+  int domain_version = dit->second;
+
+  for (const auto& attr : node.attribute()) {
+    check_attribute(attr, ctx, lex_ctx);
+  }
+
+  OpSchema* schema = nullptr;
+  std::string op_type = node.op_type();
+  std::string domain = node.domain();
+  int maxInclusiveVersion = domain_version;
+  if (map_.count(op_type) && map_[op_type].count(domain)) {
+    auto pos = map_[op_type][domain].lower_bound(maxInclusiveVersion);
+    if (map_[op_type][domain].begin() == pos &&
+        pos->first > maxInclusiveVersion) {
+      // All versions are greater than specified version.
+      fail_check(
+          "No op import for function node '" + op_type + " in op " +
+          schema->Name());
+    }
+    if (map_[op_type][domain].end() == pos ||
+        pos->first > maxInclusiveVersion) {
+      // All versions are less than specified version, or,
+      // The <pos> version is greater than specified version.
+      pos--;
+    }
+    // Schema with exact version as specified one exists.
+    schema = &(pos->second);
+  } else {
+    fail_check(
+        "No op import for function node '" + op_type + " in op " +
+        schema->Name());
+  }
+  verify_node_schema(node, schema, domain_version);
+
+  // Type constraints check between op and function body.
```

Comment:
I suggest removing this check from the current PR. There are a couple of issues here. First, the "type strings" being compared here may be type-variables, and not types. Second, this check is incomplete, since it does not do type inference for all nodes, and checks only direct references to function parameters.